### ISSUE TITLE
[Question] Question about the punctuation normlization in prepare-wmt14en2de.sh

### ISSUE THREAD

**Issue by ZhaoQianfeng** [pytorch/fairseq#3071]
**Question in short:** Why in prepare-wmt14en2de.sh, we normlise the punctuation and remove non-print char of train/valid set but not for test set?

Here is the code snippet of prepare-wmt14en2de.sh:
For train/valid  set(valid set is divided from train set afterwards):
```
echo "pre-processing train data..."
for l in $src $tgt; do
    rm $tmp/train.tags.$lang.tok.$l
    for f in "${CORPORA[@]}"; do
        cat $orig/$f.$l | \
            perl $NORM_PUNC $l | \
            perl $REM_NON_PRINT_CHAR | \
            perl $TOKENIZER -threads 8 -a -l $l >> $tmp/train.tags.$lang.tok.$l
    done
done
```
For test set:
```
echo "pre-processing test data..."
for l in $src $tgt; do
    if [ "$l" == "$src" ]; then
        t="src"
    else
        t="ref"
    fi
    grep '<seg id' $orig/test-full/newstest2014-deen-$t.$l.sgm | \
        sed -e 's/<seg id="[0-9]*">\s*//g' | \
        sed -e 's/\s*<\/seg>\s*//g' | \
        sed -e "s/\’/\'/g" | \
    perl $TOKENIZER -threads 8 -a -l $l > $tmp/test.$l
    echo ""
done
```
We dont do $NORM_PUNC and $REM_NON_PRINT_CHAR for test set.

After applying fairseq-preprocess, the preprocess.log is:
```
[en] Dictionary: 33616 types
[en] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/train.en: 3896364 sents, 116716790 tokens, 0.0% replaced by <unk>
[en] Dictionary: 33616 types
[en] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/valid.en: 39388 sents, 1179870 tokens, 0.0022% replaced by <unk>
[en] Dictionary: 33616 types
[en] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/test.en: 3003 sents, 82969 tokens, 0.00121% replaced by <unk>
[de] Dictionary: 34888 types
[de] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/train.de: 3896364 sents, 120376967 tokens, 0.0% replaced by <unk>
[de] Dictionary: 34888 types
[de] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/valid.de: 39388 sents, 1218494 tokens, 0.000903% replaced by <unk>
[de] Dictionary: 34888 types
[de] /home/user_data_182b/zhaoqf/fairseq/examples/translation/wmt14_en_de_no_normed/test.de: 3003 sents, 87323 tokens, 0.879% replaced by <unk>
Wrote preprocessed data to ../../../data-bin/wmt14_en_de_none_normed
```

Notice that in test.de, 0.879% tokens are replaced by `<UNK>`, it's so high that when I calculate BLEU score, many tokens of reference are `<UNK>`, and result in low BLEU score than valid set. For example,  valid set 29.3 but  test set 26.3.

But if i apply punctuation normlization+remove non-print char both on train/valid and test set, test.de  `<UNK> ` replacement proportion is 0.
```
[en] examples/translation/wmt14_en_de/valid.en: 39388 sents, 1179870 tokens, 0.0022% replaced by <unk>
[en] Dictionary: 33616 types
[en] examples/translation/wmt14_en_de/test.en: 3003 sents, 82968 tokens, 0.00121% replaced by <unk>
[de] Dictionary: 34888 types
[de] examples/translation/wmt14_en_de/train.de: 3896364 sents, 120376967 tokens, 0.0% replaced by <unk>
[de] Dictionary: 34888 types
[de] examples/translation/wmt14_en_de/valid.de: 39388 sents, 1218494 tokens, 0.000903% replaced by <unk>
[de] Dictionary: 34888 types
[de] examples/translation/wmt14_en_de/test.de: 3003 sents, 87321 tokens, 0.0% replaced by <unk>
```
and BLEU score of test set is 27.3, higher than 26.3 
Other settings are same for two experiments, followed the advice of [How to reproduce the result of WMT14 en-de on transformer BASE model?  issues/346](https://github.com/pytorch/fairseq/issues/346)

So I want to know is there any reason or convention that we shouldn't do punctuation normalization on test set? Looking forward to someone's reply :)

**Comment by myleott** [comment]
It's generally discouraged to modify the test set, but it's fine to make any preprocessing modifications that you want to the train/valid set. Of course, preprocessing the train set will cause a mismatch between train/test, so you have to balance the tradeoffs carefully.
